x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/regex/RegexMatch.java

```diff
@@ -0,0 +1,93 @@
+import org.elasticsearch.xpack.esql.core.expression.predicate.regex.AbstractStringPattern;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.evaluator.mapper.EvaluatorMapper;
+import org.elasticsearch.xpack.esql.expression.function.scalar.string.AutomataMatch;
+import org.elasticsearch.xpack.esql.optimizer.rules.physical.local.LucenePushdownPredicates;
+
+import java.io.IOException;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.DEFAULT;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isString;
+
+abstract class RegexMatch<P extends AbstractStringPattern> extends org.elasticsearch.xpack.esql.core.expression.predicate.regex.RegexMatch<
+    P> implements EvaluatorMapper, TranslationAware.SingleValueTranslationAware {
+
+    abstract String name();
+
+    RegexMatch(Source source, Expression field, P pattern, boolean caseInsensitive) {
+        super(source, field, pattern, caseInsensitive);
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        return isString(field(), sourceText(), DEFAULT);
+    }
+
+    @Override
+    public Boolean fold(FoldContext ctx) {
+        return (Boolean) EvaluatorMapper.super.fold(source(), ctx);
+    }
+
+    @Override
+    public EvalOperator.ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvaluator) {
+        return AutomataMatch.toEvaluator(
+            source(),
+            toEvaluator.apply(field()),
+            // The empty pattern will accept the empty string
+            pattern().pattern().isEmpty() ? Automata.makeEmptyString() : pattern().createAutomaton(caseInsensitive())
+        );
+    }
+
+    @Override
+    public Translatable translatable(LucenePushdownPredicates pushdownPredicates) {
+        return pushdownPredicates.isPushableFieldAttribute(field()) ? Translatable.YES : Translatable.NO;
+    }
+
+    @Override
+    public Expression singleValueField() {
+        return field();
+    }
+
+    @Override
+    public String nodeString() {
+        return name() + "(" + field().nodeString() + ", \"" + pattern().pattern() + "\", " + caseInsensitive() + ")";
+    }
+
+    void serializeCaseInsensitivity(StreamOutput out) throws IOException {
+        if (out.getTransportVersion().before(TransportVersions.ESQL_REGEX_MATCH_WITH_CASE_INSENSITIVITY)) {
+            if (caseInsensitive()) {
+                // The plan has been optimized to run a case-insensitive match, which the remote peer cannot be notified of. Simply avoiding
+                // the serialization of the boolean would result in wrong results.
+                throw new EsqlIllegalArgumentException(
+                    name() + " with case insensitivity is not supported in peer node's version [{}]. Upgrade to version [{}] or newer.",
+                    out.getTransportVersion(),
+                    TransportVersions.ESQL_REGEX_MATCH_WITH_CASE_INSENSITIVITY
+                );
```

Comment:
This should never happen, but I like the idea of throwing an exception here, as a defense from possible future mistakes.